File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        HashMap<Integer,Integer> map=new HashMap<>();
        map.put(0,1);
        int pref=0;
        int cnt=0;
        for(int i=0;i<nums.length;i++){
            pref+=nums[i];
            int extra=pref-k;
            if(map.containsKey(extra)){
                 cnt+=map.get(extra);  
                // map[extra]++;   
                       }
        
        map.put(pref,map.getOrDefault(pref,0)+1);
        }
    return cnt;
}
}

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int,int> mpp;  
        mpp[0] = 1;   // base case: subarray starting from index 0

        int preSum = 0, cnt = 0;

        for (int i=0; i<nums.size(); i++) {
            preSum += nums[i];                  // update prefix sum
            int remove = preSum - k;           // find target prefix
            cnt += mpp[remove];                // add how many subarrays
            mpp[preSum]++;                  // record this prefix sum
        }
    return cnt;
    }
};